src/Microsoft.AspNet.Server.Kestrel/Http/ListenerSecondary.cs

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Server.Kestrel.Networking;
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public class ListenerSecondary : ListenerContext, IDisposable
+    {
+        UvPipeHandle DispatchPipe { get; set; }
+
+        public ListenerSecondary(IMemoryPool memory)
+        {
+            Memory = memory;
+        }
+
+        public Task StartAsync(
```

Comment:
I think I have convinced myself, that this code does not have a resource leak in the exceptional cases, but it's really, really hard to get to that conclusion. I know it's asynchronous code, but is there any way to streamline the control flow?